extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/SQLCatalogManager.java

```diff
@@ -0,0 +1,777 @@
+import org.apache.druid.catalog.model.TableSpec;
+import org.apache.druid.catalog.storage.MetadataStorageManager;
+import org.apache.druid.catalog.sync.CatalogUpdateListener;
+import org.apache.druid.catalog.sync.UpdateEvent;
+import org.apache.druid.catalog.sync.UpdateEvent.EventType;
+import org.apache.druid.guice.ManageLifecycle;
+import org.apache.druid.java.util.common.ISE;
+import org.apache.druid.java.util.common.StringUtils;
+import org.apache.druid.java.util.common.jackson.JacksonUtils;
+import org.apache.druid.java.util.common.lifecycle.LifecycleStart;
+import org.apache.druid.metadata.SQLMetadataConnector;
+import org.skife.jdbi.v2.Handle;
+import org.skife.jdbi.v2.IDBI;
+import org.skife.jdbi.v2.Query;
+import org.skife.jdbi.v2.ResultIterator;
+import org.skife.jdbi.v2.Update;
+import org.skife.jdbi.v2.exceptions.CallbackFailedException;
+import org.skife.jdbi.v2.exceptions.UnableToExecuteStatementException;
+import org.skife.jdbi.v2.tweak.HandleCallback;
+
+import java.io.IOException;
+import java.util.Deque;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentLinkedDeque;
+
+@ManageLifecycle
+public class SQLCatalogManager implements CatalogManager
+{
+  public static final String TABLES_TABLE = "tableDefs";
+
+  private final MetadataStorageManager metastoreManager;
+  private final SQLMetadataConnector connector;
+  private final ObjectMapper jsonMapper;
+  private final IDBI dbi;
+  private final String tableName;
+  private final Deque<CatalogUpdateListener> listeners = new ConcurrentLinkedDeque<>();
+
+  @Inject
+  public SQLCatalogManager(MetadataStorageManager metastoreManager)
+  {
+    if (!metastoreManager.isSql()) {
+      throw new ISE("SQLCatalogManager only works with SQL based metadata store at this time");
+    }
+    this.metastoreManager = metastoreManager;
+    this.connector = metastoreManager.sqlConnector();
+    this.dbi = connector.getDBI();
+    this.jsonMapper = metastoreManager.jsonMapper();
+    this.tableName = getTableDefnTable();
+  }
+
+  @Override
+  @LifecycleStart
+  public void start()
+  {
+    createTableDefnTable();
+  }
+
+  public static final String CREATE_TABLE =
+      "CREATE TABLE %s (\n" +
+      "  schemaName VARCHAR(255) NOT NULL,\n" +
+      "  name VARCHAR(255) NOT NULL,\n" +
+      "  creationTime BIGINT NOT NULL,\n" +
+      "  updateTime BIGINT NOT NULL,\n" +
+      "  state CHAR(1) NOT NULL,\n" +
+      "  tableType VARCHAR(20) NOT NULL,\n" +
+      "  properties %s,\n" +
+      "  columns %s,\n" +
+      "  PRIMARY KEY(schemaName, name)\n" +
+      ")";
+
+  // TODO: Move to SqlMetadataConnector
+  public void createTableDefnTable()
+  {
+    if (!metastoreManager.config().isCreateTables()) {
+      return;
+    }
+    connector.createTable(
+        tableName,
+        ImmutableList.of(
+            StringUtils.format(
+                CREATE_TABLE,
+                tableName,
+                connector.getPayloadType(),
```

Comment:
BLOB type has max size of 64KB, I feel that we can easily overrun this for a table with thousands of columns. `MEDIUMBLOB` may be better